### ISSUE TITLE
Use applied ts instead of resolved ts for stale read

### ISSUE THREAD

**Issue by hihihuhu** [tikv/tikv#13930]
### Is your feature request related to a problem? Please describe:
Currently TiKV uses resolved ts to determine if a stale read is safe or not, however, when there are pending transactions, the stale read is considered not safe even if the replica has enough data. And for worse, when TiDB retries it against leader instead of follower, it will still fail as long as the pending transactions are not concluded.

### Describe the feature you'd like:
Applied ts could be more accurate to determine if a stale read is safe or not.

### Describe alternatives you've considered:
have another request mode that just returns whatever latest value available without staleness check, this mode could work better in terms of performance and well suit for our use cases, however, it is not clear if it is a thing TiDB want to support.

### Teachability, Documentation, Adoption, Migration Strategy:
Stale read is a very useful feature for multi region setup, even in a single region setup, stale read against follower could save cross availability zone network cost.

**Comment by tonyxuqqi** [comment]
PTAL @BusyJay

**Comment by BusyJay** [comment]
Can you explain more on the definition of "applied ts"?

**Comment by hihihuhu** [comment]
yeah, sure, it may not be a concept currently available in tikv but i mean to associate a safe read timestamp, probably should be just current timestamp, with each raft entry, then applied ts is basically the timestamp associated with the latest applied raft entry

**Comment by BusyJay** [comment]
We have considered applied ts when stale read is designed. We choose resolved ts in the end because applied ts needs replication and extra round of communication to advance to new ts, and it is very similar to resolved ts. We try to avoid introducing and maintaining new duplicated concepts, which will add complexity to most features and optimization.

For the shortcoming of resolved ts, we need to use other mechanism guarantee its freshness. For example:
1. We need to resolve lock actively if stale read is critical (but back in the time when we design stale read, the goal is to guarantee tens of seconds or event minutes);
2. Advance min-commit-ts for those long running transactions, this part is done as far as I know.

**Comment by hihihuhu** [comment]
i see, thanks for the explain! could you help to share some information about item 2, does it mean the long running transactions doesn't block stale read?

**Comment by hihihuhu** [comment]
thanks for the reply, however, i am still a bit confused, does a long running transaction, say runs for 10 minutes, block the stale read or not?

**Comment by cfzjywxk** [comment]
If the lock of the long-running transaction could be resolved the read would not be blocked, usually this is not a problem except in a few extreme cases.

**Comment by hihihuhu** [comment]
just to confirm i understand it, let's say there is a transactions running for minutes due legit reason (perhaps calling a few 3rd party api). then all of the stale read, regardless which table is read, will be at least a few minute stale. is my understanding correct here? thanks!

**Comment by hihihuhu** [comment]
another part of the concern is, right now, tidb has the logic to retry the stale read against leader in failure, but the leader will reject the request based on resolved ts as well.

**Comment by cfzjywxk** [comment]
> just to confirm i understand it, let's say there is a transactions running for minutes due legit reason (perhaps calling a few 3rd party api). then all of the stale read, regardless which table is read, will be at least a few minute stale. is my understanding correct here? thanks!
@hihihuhu 
If the transaction is still in the execution phase, the stale read would not be affected which means it should not be blocked and the staleness is not affected too.

If the transaction is in committing phase(2PC), if the `prewrite lock` is not committed yet and it's encountered by the read operation, the reader must try to resolve it first. Usually, the committing phase would not take too long so the stale read would not be affected too. Except in a few extreme cases the write duration in tikv is abnormal the read may be blocked for some time.

So generally speaking, the stale read would not be affected by long-running transactions.

**Comment by tonyxuqqi** [comment]
> > just to confirm i understand it, let's say there is a transactions running for minutes due legit reason (perhaps calling a few 3rd party api). then all of the stale read, regardless which table is read, will be at least a few minute stale. is my understanding correct here? thanks!
> 
> @hihihuhu
> 
> If the transaction is still in the execution phase, the stale read would not be affected which means it should not be blocked and the staleness is not affected too.
> 
> If the transaction is in committing phase(2PC), if the `prewrite lock` is not committed yet and it's encountered by the read operation, the reader must try to resolve it first. Usually, the committing phase would not take too long so the stale read would not be affected too. Except in a few extreme cases the write duration in tikv is abnormal the read may be blocked for some time.
> 
> So generally speaking, the stale read would not be affected by long-running transactions.
Maybe it's better to explain what's "resolve lock" mean? @cfzjywxk  Does it mean either to commit it or to abort it?

**Comment by cfzjywxk** [comment]
Yes, "resolve lock" is to commit or abort an encountered lock for concurrent read and write operations.

Usually, the stable read blocked by a large transaction is not related to the `resolve lock` processing, it's because the `resolved-ts` used by stale read is calculated by `start_ts` of transactions. If a large transaction could not be committed in time, its `start_ts` may block the update of `resolved_ts`, so the stale read could be rejected.
